front/pages/api/w/[wId]/files/[fileId].ts

```diff
@@ -0,0 +1,251 @@
+        ? (req.query.action as Action)
+        : "download";
+
+      // TODO(2024-07-01 flav) Expose the different versions of the file.
+      if (action === "view") {
+        const readStream = await fileRes.getReadStream(auth, "original");
+        readStream.on("error", () => {
+          return apiError(req, res, {
+            status_code: 404,
+            api_error: {
+              type: "file_not_found",
+              message: "File not found.",
+            },
+          });
+        });
+
+        readStream.pipe(res);
+        return;
+      }
+
+      // Redirect to a signed URL.
+      const url = await fileRes.getSignedUrlForDownload(auth, "original");
+
+      res.redirect(url);
+      return;
+    }
+
+    case "DELETE": {
+      const deleteRes = await fileRes.delete(auth);
+      if (deleteRes.isErr()) {
+        return apiError(req, res, {
+          status_code: 400,
+          api_error: {
+            type: "invalid_request_error",
+            message: "Failed to delete the file.",
+          },
+        });
+      }
+
+      res.status(204).end();
+      return;
+    }
+
+    case "POST": {
+      if (fileRes.isReady || fileRes.isFailed) {
+        return apiError(req, res, {
+          status_code: 400,
+          api_error: {
+            type: "invalid_request_error",
+            message:
+              "The file has already been uploaded or the upload has failed.",
+          },
+        });
+      }
+
+      if (
+        fileRes.createdAt.getTime() + UPLOAD_DELAY_AFTER_CREATION_MS <
+        Date.now()
+      ) {
+        await fileRes.markAsFailed();
+
+        return apiError(req, res, {
+          status_code: 400,
+          api_error: {
+            type: "invalid_request_error",
+            message: "File upload has expired. Create a new file.",
+          },
+        });
+      }
+
+      try {
+        const form = new IncomingForm({
+          // Stream the uploaded document to the cloud storage.
+          fileWriteStreamHandler: () => {
+            return fileRes.getWriteStream(auth, "original");
+          },
+
+          // Support only one file upload.
+          maxFiles: 1,
+
+          // Validate the file size.
+          maxFileSize: fileRes.fileSize,
+
+          // Ensure the file is of the correct type.
+          filter: function (part) {
+            if (part.mimetype !== fileRes.contentType) {
+              return false;
+            }
+
+            return true;
+          },
+        });
+        const [, files] = await form.parse(req);
+
+        const maybeFiles = files.file;
+
+        if (!maybeFiles || maybeFiles.length === 0) {
+          return apiError(req, res, {
+            status_code: 400,
+            api_error: {
+              type: "file_type_not_supported",
+              message: "File is not supported.",
+            },
+          });
+        }
+
+        const preProcessingRes = await maybeApplyPreProcessing(auth, fileRes);
```

Comment:
Not a big fine of "...Res" for resources. Should be just "file" no since ...Res is already used everywhere for results

Reply:
Agree but file was taken by formiddable, will update.